From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Rudi Heitbaum <rudi@heitbaum.com>
Date: Mon, 10 Aug 2026 00:00:00 +0000
Subject: [PATCH] nfs_v4: use void cast to allow compile for arm32

When compiling for arm32 the error: cast increases required alignment
of target type occurs. buf is a char pointer, which is only aligned to
one byte, so casting it to a wider type is diagnosed on targets that do
not permit unaligned access.

lib/nfs_v4.c: In function 'nfs_parse_attributes':
lib/nfs_v4.c:526:24: error: cast increases required alignment of target type [-Werror=cast-align]
  526 |         st->nfs_dev = ((uint64_t *)buf)[0] ^ ((uint64_t *)buf)[1];
      |                        ^
lib/nfs_v4.c:526:47: error: cast increases required alignment of target type [-Werror=cast-align]
  526 |         st->nfs_dev = ((uint64_t *)buf)[0] ^ ((uint64_t *)buf)[1];
      |                                               ^
lib/nfs_v4.c:603:42: error: cast increases required alignment of target type [-Werror=cast-align]
  603 |         st->nfs_rdev = specdata4_to_rdev((uint32_t *)buf);
      |                                          ^

Cast via (void *) as every other read of this same buf in this function
already does, for example the neighbouring size, inode and space used
attributes. The attribute buffer is 4 byte aligned, and the fsid is only
folded down to a device number rather than used as two 64 bit values, so
the reads are safe.

Signed-off-by: Rudi Heitbaum <rudi@heitbaum.com>
---
 lib/nfs_v4.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/lib/nfs_v4.c b/lib/nfs_v4.c
index f6f0149..a0554f6 100644
--- a/lib/nfs_v4.c
+++ b/lib/nfs_v4.c
@@ -523,7 +523,7 @@ nfs_parse_attributes(struct nfs_context *nfs, struct nfs4_cb_data *data,
         len -= 8;
         /* FSID */
         CHECK_GETATTR_BUF_SPACE(len, 16);
-        st->nfs_dev = ((uint64_t *)buf)[0] ^ ((uint64_t *)buf)[1];
+        st->nfs_dev = ((uint64_t *)(void *)buf)[0] ^ ((uint64_t *)(void *)buf)[1];
         buf += 16;
         len -= 16;
         /* Inode */
@@ -600,7 +600,7 @@ nfs_parse_attributes(struct nfs_context *nfs, struct nfs4_cb_data *data,
         len -= pad;
         /* raw device */
         CHECK_GETATTR_BUF_SPACE(len, 8);
-        st->nfs_rdev = specdata4_to_rdev((uint32_t *)buf);
+        st->nfs_rdev = specdata4_to_rdev((uint32_t *)(void *)buf);
         buf += 8;
         len -= 8;
         /* Space Used */
